### ISSUE TITLE
FM-00 — Keystone gate: in-process legacy→v1 converter reproducing the tiny-anchor goldens

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1565]
**Depends on:** ARCH-4 (#1563), BKD-1a (#1564) · **Blocks:** PAR-1 (#1572), FM-1 (#1586), FT-3 (#1582)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** This is the gate at the end of the Phase 2 keystone slice — the first time a *real trained legacy model* runs through the complete v1 stack (CORE-4 (#1558) basis → BKD-1 (#1560) reference backend → ARCH-2 (#1561) backbone → ARCH-3 (#1562) output layer → ARCH-4 (#1563) derivative engine) and must reproduce committed numbers end to end, not just op-level parity. The inputs are the two P0-1 tiny anchors (a `ScaleShiftMACE` and a plain `MACE`; fp64, 2 interactions, small irreps, 3 species, ZBL pair repulsion ON, each <1 MB, committed under `tests/golden/models/` with their training recipes) and their committed E/F/stress references `tests/golden/references/tiny_scaleshift_e3nn_cpu_fp64.json` / `tiny_mace_e3nn_cpu_fp64.json`. The converter imports the in-tree, byte-frozen `mace/` package **in the same process** — no separate venv (that packaging boundary is FM-1 (#1586)) — extracts the anchor to the neutral checkpoint format, loads it into the v1 architecture through the standard v1 load path, and proves parity twice: against the committed goldens, and against the live legacy model evaluated in the same process.

The anchors are **full-basis, nested-parametrization** artifacts, like every published foundation model, and getting there in develop takes a three-way disagreement no single line states: the CLI default is `--use_reduced_cg=False` (`mace/tools/arg_parser.py:240-244`), the model-class default is `True` (`mace/modules/models.py:67`), the converters assume `True` when a checkpoint config omits the key (`mace/cli/convert_e3nn_cueq.py:241`), and `SymmetricContractionWrapper.__new__` forces it back to `False` on a host without `cuequivariance` (`mace/modules/wrapper_ops.py:428`) — with `mace/tools/scripts_utils.py:293-295` recording only whatever the model ended up with. v1 stores only the reduced Clebsch-Gordan basis, so the converter owns the first real execution of the full→reduced weight projection, and it must read the *recorded* basis rather than re-deriving it from availability. Legacy checkpoints are full pickles (hence `TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD=1` in `mace/__init__.py`); the neutral format this ticket writes is the pickle-free replacement.

**Interface & constraints:**

1. **Placement / import rules.** `packages/` never imports `mace/`. The extraction side (which imports frozen legacy) therefore lives on the dual-import allowlist, under `tests/parity/` (e.g. `tests/parity/fm00_convert.py`). The import side is **not special code**: the converter must emit a file that the standard v1 `mace_torch.serialization.load_checkpoint` reads with no converter-only branch — that is the proof the neutral format is real.
2. **Neutral format** — canonical tensors in **safetensors + a JSON sidecar** (ModelConfig + per-op canonical spec), **no pickle, and never a backend `state_dict`**. **This ticket also implements the reader**, `mace_torch.serialization.load_checkpoint`: parse the JSON sidecar into descriptors, materialize the safetensors tensors through each op's `load_canonical`, and install the fp64 `e0_table` and scale/shift buffers. It is ordinary v1 code, not converter-only code — FM-00 (#1565) is simply the first ticket that needs to read a checkpoint, and everything downstream (TRN-4 (#1578) resume, DEP-1 (#1583), FM-1 (#1586)) loads through this same function rather than growing its own. Canonical layout is the one BKD-1a (#1564) pinned: `mul_ir` layout with legacy e3nn weight ordering for `linear`/`fully_connected_tp`, bias as a separate first-class tensor; the symmetric-contraction weights as one flat `[Z, A, mul]` tensor over the **reduced** Clebsch-Gordan basis in the CORE-4 (#1558) path order and per-path normalization; the radial-MLP weights as plain dense tensors sized `weight_numel` (backend-independent).
3. **Full→reduced projection.** The symmetric contraction is linear in its weights and the full basis's extra directions are pure gauge: measured at fp64 on published artifacts, the full basis has rank equal to the reduced parameter count and the least-squares projection is exact to ~1e-13 (MACE-MP-0 small: 28→13 paths, max err 3.2e-14; MACE-OFF23 medium: 86→29 paths, 8.9e-14) and **unique** in that direction. Consequences that bind this ticket: (a) the conversion is **function-preserving, not bit-exact in weight space** — parity is gated on E/F/stress outputs, never on weight equality; (b) the projection is one small `[A_full, A_reduced]`-shaped matrix per contraction, reused across elements and channels (weights are indexed `[element, path, channel]`); (c) reduced→full is under-determined and is **refused** here (full-basis export for external reimplementations is DEP-3 (#1595)).
4. **Independent Clebsch-Gordan pin (the gate's hard clause).** The projection is computed with CORE-4 (#1558)'s native `full ↔ reduced` machinery — but it must be **verified against an independent Clebsch-Gordan derivation**, never against CORE-4 (#1558)'s own vendored basis, never against legacy's vendored tensors in `mace/tools/cg.py`, and never against `mace/tools/cg_cueq_tools.py::symmetric_contraction_proj`, which is what the legacy converter itself calls (`mace/cli/convert_e3nn_cueq.py:11,144`). Concretely: a standalone CPU derivation (first-principles numpy/sympy, no runtime library) recomputes the basis for the anchor's `(irreps_in, irreps_out, correlation)` and the ticket asserts the projection agrees with it at fp64. A bug in CORE-4 (#1558) must fail this gate, not self-validate through it.
5. **E0 / scale-shift mapping** is a 1:1 buffer transfer, no refit: legacy `atomic_energies_fn.atomic_energies` (`mace/modules/blocks.py:364`, `AtomicEnergiesBlock`) → the v1 fp64 `e0_table` with its provenance stamped `"legacy-converted"`; legacy `scale_shift.{scale,shift}` (`mace/modules/blocks.py:1942`, `ScaleShiftBlock`) → the v1 output-layer buffers. This transfer introduces zero numerical error; parity of the E0 pathway is exact at fp64, and the ARCH-3 (#1562) two-reduction form (`Σ node_e0` and `Σ node_interaction` reduced separately, one graph-level add) is what makes that reachable.
6. **Both anchors are mandatory.** ZBL pair repulsion is applied *scaled inside* `ScaleShiftMACE` (`node_es_list = [pair_node_energy]`, `mace/modules/models.py:545`) but *added outside* the scale in plain `MACE` (`energies = [e0, pair_energy]`, `mace/modules/models.py:361-362`); only converting both turns that divergence into a numeric gate. The mapping of the ZBL term must be correct per model class.
7. **Two-sided verification, one tolerance source.** (a) Committed goldens: E/F/stress on all P0-1 fixtures at the fp64 row of the tolerance table in `tests/golden/harness.py` (1e-6 eV / 1e-6 eV/Å; single source of truth, edit-locked — a tolerance change is its own reviewed PR, never part of this ticket). (b) Live legacy in the same process: evaluate the frozen legacy anchor and the converted v1 model side by side and diff via the framework-agnostic `harness.py` snapshot machinery (numpy/ase/json only — it imports neither stack). Since the only non-exact step is the ~1e-13 basis projection, observed deviations should sit orders of magnitude below the 1e-6 gate — treat anything near the gate as a bug, not headroom. Restore any global state the legacy import mutates (`TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD`, default dtype, RNG) after the legacy pass; the full snapshot/restore harness is PAR-1 (#1572)'s job, but FM-00 (#1565) must not leak legacy process state into the v1 evaluation.

**Task:**
1. `tests/parity/fm00_convert.py` (dual-import allowlist): load a P0-1 anchor through frozen `mace/`, walk its modules, and extract every weight to the canonical layout — `mul_ir` linears with separate bias, radial-MLP dense weights, the flat `[Z, A, mul]` symmetric-contraction tensor (still full-basis at this point), E0 table, scale/shift, and the JSON model config (irreps, correlation, cutoff, ZBL on, model class, `clebsch_gordan_basis` as *recorded* by `scripts_utils.extract_config_mace_model`, not as re-derived).
2. Apply CORE-4 (#1558)'s full→reduced projection to the symmetric-contraction weights; write the neutral artifact (safetensors + JSON, `clebsch_gordan_basis: "reduced"`).
3. Add the independent-derivation check of constraint 4: recompute the reduced Clebsch-Gordan basis for the anchor's configuration from first principles in a standalone module and assert the projection used in step 2 agrees at fp64.
4. Load the artifact through the standard `mace_torch.serialization.load_checkpoint` into the v1 architecture (reference backend) — no special-case code path.
5. Parity tests `tests/parity/test_fm00_keystone.py`: for **both** anchors, on all P0-1 fixtures, (a) reproduce the committed E/F/stress goldens at the fp64 row; (b) match the live legacy model evaluated in the same process; (c) reproduce the P0-6 single-step gradient golden (`tests/golden/references/tiny_{scaleshift,mace}_train_step_grad_fp64.json`): with the converted weights (function-identical, init-agnostic), one forward+backward of the energy+forces loss on the tiny fixture matches the committed `d(loss)/d(params)` — this is the check that the conversion preserved the `create_graph=True` second-derivative path force training depends on.
6. Wire the tests into CI so the keystone gate is a green/red signal on `develop`.

**Out of scope:** published foundation models (MACE-MP/OFF/MDP conversion is FM-2 (#1587); MACE-Polar is ELEC-2 (#1592)); the production two-step converter and the pinned-venv end-user packaging path (FM-1 (#1586)); reduced→full export (DEP-3 (#1595)); the continuous parity harness with full process-state snapshot/restore and the 68 black-box test functions under `--engine v1` (PAR-1 (#1572)); regenerating any golden (goldens are edit-locked).

**Acceptance criteria:**
- [ ] Both P0-1 anchors convert in-process (frozen in-tree `mace/`, no separate venv) to a safetensors+JSON neutral artifact containing only canonical tensors — no pickle, no backend `state_dict` keys.
- [ ] The artifact loads through the standard v1 `load_checkpoint` (reference backend) with no converter-specific code path.
- [ ] Converted models reproduce the committed E/F/stress goldens on all P0-1 fixtures at the fp64 tolerance row (1e-6 eV / 1e-6 eV/Å) for **both** model classes (ZBL scaled-inside vs added-outside both exercised).
- [ ] Converted models match the live legacy anchors evaluated in the same process, compared through the framework-agnostic golden harness.
- [ ] The full→reduced projection is verified against an independent first-principles Clebsch-Gordan derivation — not CORE-4 (#1558)'s own basis, not `mace/tools/cg.py`, not `cg_cueq_tools.symmetric_contraction_proj`; no parity assertion anywhere gates on weight equality.
- [ ] The converter reads the recorded `use_reduced_cg` from the checkpoint config and refuses (rather than guessing) when the key is absent and the weight shapes are ambiguous — develop's three defaults never leak into v1.
- [ ] E0s and scale/shift arrive as a 1:1 buffer transfer: fp64 `e0_table` with provenance `"legacy-converted"`, scale/shift in the output-layer buffers; the E0 pathway contributes zero conversion error.
- [ ] The P0-6 single-step `d(loss)/d(params)` gradient golden is reproduced by the converted models (one fwd+bwd of an energy+forces loss, fp64 CPU).
- [ ] The converter refuses a reduced→full request.

**Verify:**
```bash
python -m pytest tests/parity/test_fm00_keystone.py -v          # goldens + live legacy + gradient golden
python -m pytest tests/parity/test_fm00_independent_basis.py -v # projection vs independent derivation
python -m pytest tests/golden/ -v                               # goldens themselves untouched and green
```
**Review focus:** the full→reduced projection and its independent-derivation pin (needs group-theory review — this is the step that can be silently wrong while looking plausible), the E0/scale-shift buffer mapping, and that no assertion compares weights instead of outputs.

---
